extras/indexing/src/main/java/org/apache/rya/indexing/entity/EntityIndexOptimizer.java

```diff
@@ -0,0 +1,73 @@
+package org.apache.rya.indexing.entity;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import org.apache.log4j.Logger;
+import org.apache.rya.indexing.entity.model.Entity;
+import org.apache.rya.indexing.entity.model.Type;
+import org.apache.rya.indexing.entity.query.EntityQueryNode;
+import org.apache.rya.indexing.entity.storage.EntityStorage;
+import org.apache.rya.indexing.entity.storage.TypeStorage;
+import org.apache.rya.indexing.external.matching.AbstractExternalSetOptimizer;
+import org.apache.rya.indexing.external.matching.ExternalSetMatcher;
+import org.apache.rya.indexing.external.matching.ExternalSetProvider;
+import org.apache.rya.indexing.external.matching.QueryNodeListRater;
+import org.apache.rya.indexing.external.matching.QuerySegment;
+import org.apache.rya.indexing.external.matching.TopOfQueryFilterRelocator;
+import org.openrdf.query.BindingSet;
+import org.openrdf.query.Dataset;
+import org.openrdf.query.algebra.TupleExpr;
+
+import com.google.common.base.Optional;
+
+/**
+ * Optimizes a query over {@link Entity}s.
+ */
+public class EntityIndexOptimizer extends AbstractExternalSetOptimizer<EntityQueryNode> {
+    private static final Logger log = Logger.getLogger(EntityIndexOptimizer.class);
+    private static final EntityExternalSetMatcherFactory MATCHER_FACTORY = new EntityExternalSetMatcherFactory();
+    private final EntityIndexSetProvider provider;
+
+    /**
+     * Creates a new {@link EntityIndexOptimizer}.
+     * @param typeStorage - The mechanism for access the various {@link Type}s. (not null)
+     * @param entityStorage - The mechanism for access to the various {@link Entity}s. (not null)
+     */
+    public EntityIndexOptimizer(final TypeStorage typeStorage, final EntityStorage entityStorage) {
```

Comment:
Unfortunately, you need to be able to create the Optimizer from  a Configuration object.  See RdfCloudTripleStoreConnection 316-335.  This is where your optimizer will be used.  You also need to add setters and getters to the MongoIndexingConfiguration object to configure Rya to add your optimizer as an additional optimizer.